### PR TITLE
General 1D & 2D Pooling layers

### PR DESCRIPTION
**Pull request by maxpumperla**
We address  #522,  #230, #768, #786 and propose a solution that should cover most of those tickets. PR includes

- Added Pooling1D and Pooling2D classes in convolutional
- MaxPooling1D/2D and MeanPooling1D/2D as derived classes implemented (instead of "pool_mode" flags)
- In particular, all existing examples change untouched (same API as before)
- Extended tests and docs

Note: I didn't include SumPooling, bc there seems to be no implementation in TF (that I know of), but I included a suggestion for Theano in comments. If wanted, we can pretty quickly implement other pooling modes later on.

### CODE REVIEW

**Comment by maxpumperla** [comment]
Sorry, had to patch - lost in translation in theano docs.

**Comment by fchollet** [comment]
This looks like a valuable addition, thanks. I'll try to review it today.

One immediate nitpick: I think `AveragePooling` would be a more appropriate name, at least that's the name that comes up in the literature.

**Comment by fchollet** [comment]
Also: could you solve the current merge conflicts?

File: keras/backend/theano_backend.py [hunk 1]
```diff
@@ -596,19 +596,32 @@ def maxpool2d(x, pool_size, strides=(1, 1), border_mode='valid',
...
-                                      ignore_border=ignore_border,
-                                      padding=padding,
-                                      mode='average_exc_pad')
+    if pool_mode == 'max':
+        pool_out = downsample.max_pool_2d(x, ds=pool_size, st=strides,
+                                          ignore_border=ignore_border,
+                                          padding=padding,
+                                          mode='max')
+    elif pool_mode == 'mean':
+        # Only seems to work with mode='ignore_borders' right now
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: 'ignore_border' (typo). Why?

File: keras/backend/theano_backend.py [hunk 2]
```diff
@@ -596,19 +596,32 @@ def maxpool2d(x, pool_size, strides=(1, 1), border_mode='valid',
...
+        pool_out = downsample.max_pool_2d(x, ds=pool_size, st=strides,
+                                          ignore_border=ignore_border,
+                                          padding=padding,
+                                          mode='average_exc_pad')
+    elif pool_mode == 'sum':
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: Either this mode should be removed (not supported by the tf backend) or code redundancy with 'max' should be eliminated. I'd recommend removing the mode.

File: keras/backend/tensorflow_backend.py [hunk 3]
```diff
@@ -575,10 +575,20 @@ def maxpool2d(x, pool_size, strides=(1, 1),
         # TH kernel shape: (depth, input_depth, rows, cols)
         # TF kernel shape: (rows, cols, input_depth, depth)
         x = tf.transpose(x, (0, 2, 3, 1))
-        x = tf.nn.max_pool(x, pool_size, strides, padding=padding)
+        if pool_mode=='max':
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: This should be refactored to eliminate code redundancy. The conditional flow should affect the reshape ops.

**Comment by maxpumperla** [comment]
Should be OK now. When you merge, I think  #768, #786  could be closed.

**Pull request merged**